add link to manpages in HTML builder

anarcat opened the pull request:
- Feature


- Create actual HTML links (`<a href>`) when the `:manpage:` role is
  used in RST markup.
- Requires a new configuration setting, but is backwards-compatible
  with older configuration files.


It seems way more useful to have the HTML documentation builder
actually link to real rendered versions of HTML manpages in its
output. That way people can click on manpages to get the full
documentation. There are a few services offering this online, so we do
not explicitly enable one by default, but the Debian manpages
repository has a lot of them.

No unit tests because I couldn't quite figure out where to fit that
in.

tk0miya commented on doc/config.rst:
```diff
@@ -295,6 +295,14 @@ General configuration
 
    .. versionadded:: 1.3
 
+.. confval:: manpages_url_pattern
```
I feel `manpages_url` is simpler. What do you think?

anarcat replied:
I don't care - but i felt that "pattern" implied it was a bit more than just a prefix...

tk0miya replied:
If you don't opposite, could you rename this please?

anarcat replied:
sure. done.

tk0miya commented on doc/config.rst:
```diff
@@ -295,6 +295,14 @@ General configuration
+.. confval:: manpages_url_pattern
+
+   A URL to cross-reference manpages directives. If this is defined to
+   ``https://manpages.debian.org/%s``, the ``:manpage:ls(1)`` role
```
For example, manpages.org uses only the name part (without section) like http://manpages.org/ls .
So I propose you to use named format arguments like `{full}`, `{name}` and `{section}`.

tk0miya replied:
In addition, it would be very useful to add a hyperlink to here onto an introduction of `:manpage:` role.

anarcat replied:
Yeah, I thought about doing a tuple - but how do you parse the manpage? Sometimes people don't put a section, sometimes it's page.section or page(section), it's not really well normalized.

But if you think that's important, i'll try to work on a regex and switch to `format`-style patterns.

anarcat replied:
okay, so I got this now:

```python
    def visit_manpage(self, node):
        # type: (nodes.Node) -> None
        self.visit_literal_emphasis(node)
        if self.manpages_url_pattern:
            manpage = ' '.join([str(x) for x in node.children
                                if isinstance(x, nodes.Text)])
            pattern = r'^(?P<full>(?P<name>.+)[\(\.](?P<section>[1-9]\w*)?\)?)$'
            r = re.match(pattern, manpage)
            if r:
                info = r.groupdict()
            else:
                info = {'full': manpage}
            node['refuri'] = self.manpages_url_pattern.format(**info)
            self.visit_reference(node)
```

but my concern is that I have to copy this over the `html5.py` builder again, and then this will be duplicated all over the place. what would be a better place to keep that functionality?

tk0miya replied:
Good work. I like it :-)

>what would be a better place to keep that functionality?

I think this is a kind of transform module.
```
# in sphinx/transform/__init__.py
class ManpageTransform(SphinxTransform):
    def apply(self):
        for node in self.document.traverse(addnodes.manpage):
            manpage = ' '.join([str(x) for x in node.children
                                if isinstance(x, nodes.Text)])
            pattern = r'^(?P<full>(?P<name>.+)[\(\.](?P<section>[1-9]\w*)?\)?)$'
            r = re.match(pattern, manpage)
            if r:
                info = r.groupdict()
            else:
                info = {'full': manpage}
            for key, value in info.items():
                node[key] = value

# => Add this transform class to the transforms list in sphinx/io.py
```

This allows you to access `name`, ` section` and `full` as attributes of nodes.
And it also make writers simple.

anarcat replied:
i see, done.

tk0miya commented on sphinx/config.py:
```diff
@@ -123,6 +123,7 @@ class Config(object):
         primary_domain = ('py', 'env', [NoneType]),
         needs_sphinx = (None, None, string_classes),
         needs_extensions = ({}, None),
+        manpages_url_pattern = (None, 'html'),
```
I think this should be effect not to only html, but also all builders (in future). For example, it would be nice if LaTeX builder also support the hyperlinks. \^So last argument should be `'env'`. It means the update of settings effects to all builders.

anarcat replied:
ah. okay. done.

tk0miya commented on sphinx/templates/quickstart/conf.py_t:
```diff
@@ -117,6 +117,9 @@ html_static_path = ['{{ dot }}static']
 
+# Cross-reference :manpage: roles to actual sites.
+#
+# manpage_url_pattern = 'https://manpages.debian.org/%s
```
To make default `conf.py` simple, please do not add new comment-block her.

anarcat replied:
sorry, what do you mean? should i add it elsewhere? or not at all? i added this because it says that should be done on top of the Config class.

tk0miya replied:
Oh, sorry. Please forget the comment of Config class. I will change it later.
Current `conf.py` contains a lot of comments. I think it confuses users. So we have a plan to reduce these comments from `conf.py`.

tk0miya replied:
Could you remove this please?

anarcat replied:
done. it wasn't clear from your comment you actually wanted it removed (or if you would remove it yourself at a later point).

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -79,6 +79,7 @@ def __init__(self, builder, *args, **kwds):
         self.highlightopts = builder.config.highlight_options
         self.highlightlinenothreshold = sys.maxsize
         self.docnames = [builder.current_docname]  # for singlehtml builder
+        self.manpages_url_pattern = getattr(builder.config, 'manpages_url_pattern')
```
After adding an entry to `sphinx.config`, you can access it always. So you don't need to use `getattr` here. 
BTW, why are you storing the value of `manpage_url_pattern` as an instance variable?

anarcat replied:
because that's the pattern used for other configuration, like `highlightopts`.

anarcat replied:
and the reason why i'm using `getattr` is so that older configuration files keep working, but maybe that's not necessary?

tk0miya replied:
Yes, it's not necessary. It is always enabled even if with old config-files, because you'd added it to `sphinx.config`.

>because that's the pattern used for other configuration, like highlightopts.

I see, either is okay to me.

anarcat replied:
okay, fixed.

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -816,10 +817,17 @@ def depart_abbreviation(self, node):
+            manpage = ' '.join([str(x) for x in node.children
+                                if isinstance(x, nodes.Text)])
+            node['refuri'] = self.manpages_url_pattern % manpage
+            self.visit_reference(node)
```
Could you tell me the result if `manpage_url_pattern` enabled?
It seems this visitor method call both `visit_literal_emphasis` and `visit_reference`. So I can't guess it.

anarcat replied:
This:

```
:manpage:`hdparm(8)`
```

... gets turned in:

```
<em class="manpage"><a class="manpage reference external" href="https://manpages.debian.org/hdparm(8)">hdparm(8)</a></em>
```

i figured i would keep the existing emphasis style.

tk0miya reviewed (changes requested):
looks good to me for your concept.
Could you add testcase for this please?

anarcat commented:
>  Could you add testcase for this please?

See, that's where I'm stuck. Where should i add it?

anarcat commented:
i re-rolled a new patch with improved pattern support and documentation. i would need help to figure out where to put those eventual tests and how to avoid code duplication.

tk0miya commented on doc/config.rst:
```diff
@@ -295,6 +295,21 @@ General configuration
+
+   A URL to cross-reference :rst:role:`manpage` directives. If this is
+   defined to ``https://manpages.debian.org/{full}``, the
+   ``:manpage:man(1)`` role will like to
```
Please use
```
:literal:`:manpage:`man(1)``
```
instead. It allows to represent role notations correctly.

anarcat replied:
that's quite a mouthful! but okay, done. :)

tk0miya commented:
>See, that's where I'm stuck. Where should i add it?

Please add testcase to `tests/test_build_html.py`.
`test_numfig_with_numbered_toctree` is a good example for you. It modifies configuration using `confoverrides` parameter of `@pytest.mark.sphinx`, and verifies the generated HTML with `@pytest.mark.parametrize` decorator.

anarcat commented:
alright, i pushed a new patch with the recommended changes. next step is for me to figure out unit tests.

anarcat commented:
i *think* i got the test suite right. i had some weird failures *elsewhere* in the test suite, but i suspect this may be my local environment that was broken... we'll see what travis says.

i'm not sure what the second argument in the xpath checks is. i left it empty, but i have tried `man(1)` and `:manpage:...` to no avail. any ideas?

anarcat commented:
whoa, and the test suite actually passes now. woot! :)

tk0miya commented on doc/config.rst:
```diff
@@ -295,6 +295,21 @@ General configuration
+
+     * ``manpage_name`` - the manpage name (``man``)
+     * ``manpage_section`` - the manpage section (``1``)
+     * ``manpage_full`` - the whole string (``man(1)``)
```
I feel `manpage_` suffix is a bit verbose. It is obvious that these variables are related with manpage, because this settings is used for URL of manpages.

anarcat replied:
sure, but internally, the `node` parameter would be ambiguous otherwise. since i just pass the whole `node` attributes to the formatter, i figured it was better to have unambiguous parameters.

now, i could expand that formatter to have an explicit dictionary with only those values, but that didn't seem very useful. i will do that, however, if you insist on changing the formatter...

tk0miya replied:
I feel `manpage_` prefix is also too verbose for node attributes. `manpage_node['name']` and `manpage_node['section']` are much simpler for me.
What do you think?

anarcat replied:
er. well i explained this earlier: i thought having a `name` parameter in the node would be ambiguous. but i don't know how the parsers work - if you're saying it's unambiguous i'm happy to fix this!

tk0miya replied:
I don't think so. Usually, we know the type of nodes when we manipulate them. For example, `ManpageLink` transform searches manpage node using `node.traverse()`.  And each `visit_*` methods of writer also know the type of node.
So I feel `manpage_name` of manpage node is verbose. We already know the node represents a reference for manpage.
I understand `any_node['name']` is ambiguous. But we would not access in such way.

In addition, I feel the format params of `manpages_url` is verbose.

anarcat replied:
understood, i'll fix this then! :)

tk0miya commented on doc/config.rst:
```diff
@@ -295,6 +295,21 @@ General configuration
+
+   This also supports manpages specified as ``man.1``.
+
+   .. versionadded:: 1.7
```
Please mention to this confval at an introduction of :manpage: role.
http://www.sphinx-doc.org/en/stable/markup/inline.html#role-manpage

anarcat replied:
done.

tk0miya replied:
This implementation only effect to HTML output. So we should mention users the limitation (or rename this to `html_manpages_url`).

anarcat replied:
added a note.

tk0miya reviewed (commented):
LGTM with nits

anarcat commented:
i believe i fixed all your remaining concerns, except the `manpage_` prefix in the node attributes and formatter. let me know if you still think that's important to change.

shimizukawa commented:
FYI. The code freeze day of 1.7 will come in a week.

> Mid of Jan: Feature freeze, Call for translation and release 1.7b1
https://github.com/sphinx-doc/sphinx/issues/4211#issuecomment-343652178

anarcat commented:
i remerged with master here to resolve conflicts, and I believe that I addressed all the concerns that were formulated here by @tk0miya - so i'm not sure what the next step is here.

tk0miya commented:
@anarcat Please check the result of Travis CI and Circle CI. It seems many tests are failed. I will review it after resolved.

tk0miya commented on sphinx/config.py:
```diff
@@ -125,6 +125,7 @@ class Config(object):
         primary_domain = ('py', 'env', [NoneType]),
         needs_sphinx = (None, None, string_classes),
         needs_extensions = ({}, None),
+        manpages_url_pattern = (None, 'env'),
```
I guess the error comes from here. This is not renamed.

anarcat replied:
fixed.

anarcat commented:
> @anarcat Please check the result of Travis CI and Circle CI. It seems many tests are failed. I will review it after resolved.

i pushed another fix and conflict resolution, let's see what CI thinks.

i'd also like to hear from you about the ambiguity question...

anarcat commented:
@tk0miya CI passes, this is ready for a final review. i think the only question is the `manage_` prefix in the node object, if it is redundant or if just having e.g. `name` there would be ambiguous.

thanks!

anarcat committed:
add link to manpages in HTML builder

It is useful to have the HTML documentation builder actually link to
real rendered versions of HTML manpages in its output. That way people
can click on manpages to get the full documentation. There are a few
services offering this online, so we do not explicitly enable one by
default, but the Debian manpages repository has a lot of the manpages
pre-rendered, so it is used as an example in the documentation.

The parsing work is done by a transformer class that parses manpage
objects and extract name/section elements. Those then can be used by
writers to cross-reference to actual sites. An implementation is done
in the two HTML writers, but could also apply to ePUB/PDF writers as
well in the future.

This is not enabled by default: the `manpages_url` configuration item
needs to be enabled to point to the chosen site. The `page`, `section`
and `path` parameters are expanded through Python string formatting in
the URL on output.

Unit tests are fairly limited, but should cover most common use-cases.

anarcat commented:
okay, so I did the following renames:

 * `manpage_name` is now `page`, to fit with the naming convention in `man(1)`
 * `manpage_section` is now `section`
 * `manpage_full` is now `path` because `full` seemed strange without context

CI seems to pass, so I think this would be ready to merge if you agree with the new naming.

Thanks for the prompt review!

tk0miya reviewed (approved):
LGTM. Thank you for update :-)

tk0miya commented:
Merged. Thank you for your contribution!

anarcat commented:
yay, thanks!
